src/module_wrap.cc

```diff
@@ -885,33 +896,32 @@ void ThrowExportsInvalid(Environment* env,
                          const URL& pjson_url,
                          const URL& base) {
   Local<String> target_string;
-  if (target->ToString(env->context()).ToLocal(&target_string)) {
-    Utf8Value target_utf8(env->isolate(), target_string);
-    std::string target_str(*target_utf8, target_utf8.length());
-    if (target->IsArray()) {
-      target_str = '[' + target_str + ']';
-    }
-    ThrowExportsInvalid(env, subpath, target_str, pjson_url, base);
+  if (target->IsObject()) {
+    target_string = v8::JSON::Stringify(env->context(), target.As<v8::Object>(),
+                                        env->empty_string()).ToLocalChecked();
+  } else {
+    target_string = target->ToString(env->context()).ToLocalChecked();
```

Comment:
Can we try to keep proper handling of failed V8 calls? :confused:

Reply:
What would help to ensure proper handling of failed V8 calls would be if I even knew what v8 even does on these failure paths, since there are zero docs on any of this stuff.

The only way I can see this failing in JS is if one had a custom `Object.prototype.toString` that throws, therefore this operation would throw.

In that case does v8 throw the JS error for the underlying trap? So that we can assume `try_catch.HasCaught` in this case?

Reply:
> The only way I can see this failing in JS is if one had a custom `Object.prototype.toString` that throws, therefore this operation would throw.

That’s right, but the issue is that *any* operation that ends up calling JS code can throw, even if the JS code can’t throw an exception by itself, in the presence of termination exceptions (i.e. “magic” uncatchable exceptions from another thread) such as the ones generated by `worker.terminate()`.

> since there are zero docs on any of this stuff.

Yeah, that’s on me. I’ll try to update the C++ `src/README.md` and the V8 docs for `TryCatch` with more info on this subject.

Reply:
Thanks for the clarification here, can you confirm that the catch will have caught in both of these cases? Because this code path is about ensuring an error is thrown in the environment (of some sort). So if it is an error throwing an error we need to be sure that will at least be propagated.

Reply:
@guybedford Yeah, a `TryCatch` will catch termination exceptions as well. :+1: